processor/memorylimiterprocessor/memorylimiter.go

```diff
@@ -149,12 +153,16 @@ func (ml *memoryLimiter) start(_ context.Context, host component.Host) error {
 		}
 	}
 
-	ml.startMonitoring()
+	ml.startOnce.Do(func() {
+		ml.startMonitoring()
+	})
 	return nil
 }
 
 func (ml *memoryLimiter) shutdown(context.Context) error {
-	ml.ticker.Stop()
+	ml.shutDownOnce.Do(func() {
+		ml.ticker.Stop()
```

Comment:
If one processor is used in multiple pipelines isn't there a chance this could be preemptive and halt functionality for remainders that have yet to shutdown? We probably want to keep an atomic or waitgroup for the last instance to call.

Reply:
I was not aware that we can reload the config dynamically to result of multiple start/shutdown; other than a unit test, how can I trigger this as a user?

@bogdandrutu thanks for pointing out the `SharedComponent` it did the trick to get rid of the sync.once for the start/shutdown, however, if I tried to the get rid of the `getMemoryLimiter` factory logic and fully count on `GetOrAdd` then it won't work well as the `SharedComponent` does not handle errors of object initialization; 
In other words, if my func looks like this... 

````
func createTracesProcessor(
	_ context.Context,
	set component.ProcessorCreateSettings,
	cfg config.Processor,
	nextConsumer consumer.Traces,
) (component.TracesProcessor, error) {
	var err error
	var ml *memoryLimiter
	mlInstance := memLimiters.GetOrAdd(cfg, func() component.Component {
		ml, err = newMemoryLimiter(set, cfg.(*Config))
		return ml
	})
	if err != nil {
		return nil, err
	}
	return processorhelper.NewTracesProcessor(
		cfg,
		nextConsumer,
		mlInstance.Unwrap().(*memoryLimiter).processTraces,
		processorhelper.WithCapabilities(processorCapabilities),
		processorhelper.WithStart(mlInstance.Start),
		processorhelper.WithShutdown(mlInstance.Shutdown))

}
```
....and `newMemoryLimiter` was supplied wrong config to cause an error, this component will still be cached by `SharedComponent`  and yes, the error check outside the `SharedComponent` scope will result of returning `nil, err` but subsequent `create*Processor`, with the same config, will pick up the cached value component; in my case, causing unit tests to fail.

The only workaround I can think of is to unwrap the object and check if it's nil, but I would loose the error message.
```
	if mlInstance.Unwrap().(*memoryLimiter) == nil {
		return nil, fmt.Errorf("memory limiter object failed to get initialized")
	}
```
Also note, that this can cause some type of leaks as those objects are cached and we'll not call shutdown on them to clear them up from . This issue exists at the moment `SharedComponent` map.

I might be missing something that does not require changing the behavior of `SharedComponent` create.

Reply:
@rmfitzpatrick 
I tried a combination of atomic and waitgroup, but that would not be 100% thread safe as I need to sync multiple steps, so I reverted to basic mutex which works fine for this usage and performance should not take a hit as the start/shutdown should not be called frequently

@bogdandrutu 
I looked at `SharedComponent` implementation and it uses sync once per instance, so this and in addition to lack of error checking, I don't think `SharedComponent` is a good fit here. 
And as I mentioned in my previous comment, in case of an error, the component will leak, that should also be addressed

@tigrannajaryan @bogdandrutu
Is it possible to call shutdown without calling start? I am checking for that and it using it in the test, but was not sure if there are extra precautions needed